File: Dynamic Programming/DP on MCM - Partition DP/Palindrome Partitioning II/recursive.java
class Solution {
    public int minCut(String s) {
        return helper(0, s) - 1; 
    }

    private int helper(int start, String s) {
        if (start == s.length()) return 0;

        int min = Integer.MAX_VALUE;
        for (int end = start; end < s.length(); end++) {
            if (isPalindrome(s, start, end)) {
                min = Math.min(min, 1 + helper(end + 1, s));
            }
        }
        return min;
    }

    private boolean isPalindrome(String s, int l, int r) {
        while (l < r) {
            if (s.charAt(l++) != s.charAt(r--)) return false;
        }
        return true;
    }
}

File: Dynamic Programming/DP on MCM - Partition DP/Palindrome Partitioning II/recursive.cpp
class Solution {
    public:
        bool isPalindrome(int l, int r, string &s){
            while(l < r){
                if(s[l] != s[r]) return false;
                l++;
                r--;
            }
            return true;
        }
    
        int solve(int i, string& s) {
            if(i >= s.size()) return 0;
            int minCost = INT_MAX;
            for(int j=i; j<s.size(); j++){
                if(isPalindrome(i, j, s)){
                    int cost = 1 + solve(j+1, s);
                    minCost = min(minCost, cost);
                }
            }
    
            return minCost;
        }
        
        int minCut(string s) {
            return solve(0, s) - 1;
        }
    };